### ISSUE TITLE
Support use of savefiles as test scenario to check engine compatibility

### ISSUE THREAD

**Issue by demario-w** [wesnoth/wesnoth#7029]
### Describe the desired feature

As test procedures are hard to write, we are sitting on tons of savefiles that contains numerous combinations of units/traits/capabilities/terrains/ToD/... where the expected behavior is known: the result stored in the savefile.

As an illustration, the way to identify the incompatible way of rounding xp according traits was by hitting OOS during replay of 1.14 saves on BFW 1.16. If that test had been run during development, the problem would have been caught before release.

If a savefile can be replayed in background mode, the run terminates and the output shows if an OOS has occurred or not, all these saves could be used as test scenarios to verify the behavior of the game engine in various situations. All problems in engine do not lead to a straightforward OOS (as the complete game state is not available in the save) but most of them will ultimately ends up in a OOS (missing attack target, difference in damage, unmatched death, target hex cannot be reached, gold not enough...). If the number of saves used as tests is large enough, the likeliness that a given problem in the engine will end up in creating an OOS in one of them is high.

Most of the available savefiles are not generated on the version of the engine where the development is done. So if this trove of saves need to be exploited to generate test scenario, there is a need that the engine is supporting replay of past version saves.

Many known saves have been identified as catching backward-incompatibilities in the engine and they would be the first targets to be turned into scenarios to test against regression. 

Being able to replay old saves also allows to exercise the engine in ways that is not possible with saves from the current version by picking saves from game scenarios that haven't been ported to the current version (which is the master branch).

**Comment by demario-w** [comment]
For the record, here is how I start a replay from a save right now:

$ wesnoth --core=core_1_14 --with-replay --validcache --load $i

where $i hold the name of the save file.
I need 4 more "clicks" (or hotkey press) to switch to the next file: "skip animation", "continuous replay" and double click on the button to close the window at the end. Can this number be reduced further?

The result of the test (not OOS) is not passed back to the calling shell at this time.

**Comment by stevecotton** [comment]
It sounds like a good feature.

The current `--unit` test mechanism in `game_launcher::single_unit_test()` has something similar, but which is only used immediately after running the unit-test scenarios. Each test is recorded with the hardcoded filename of "unit_test_replay", and then it checks that it can be replayed without OOS.

**Comment by CelticMinstrel** [comment]
Yeah, so the functionality already exists in the engine, which means it just needs to be exposed.

**Comment by demario-w** [comment]
> The current `--unit` test mechanism in `game_launcher::single_unit_test()` has something similar

The difference between replaying a unit test and a random saves is that most of the saves do not end with a victory or defeat. Would that impact the ability to return with a result at the end of the replay?

**Comment by demario-w** [comment]
I came up with the following command line to run a replay as a test:
> (./wesnoth --data-dir=. --validcache --replay_test 202111...-ladder.gz)

I am recycling the existing flag is_unit_test to trigger the behavior that I want. So that behavior will extend to all unit tests that use a replay.

The current behavior is:
- activate the skip animation per default
- start the continuous replay automatically
- in case of OOS, the standard dialog ignore/quit/save still shows
- if no winner at the end of the replay, the game still closes (to allow running them in batch)

A comment on the case where OOS occurs: while I was looking to disable the dialog, I think I like to keep it now. It implements a feature to "fallback to interactive mode in case of failure" that is helpful for investigating the OOS (the fallback is limited as the whole test is obviously running interactively).
